### PR TITLE
refactor(linter): add `LintFilter`

### PR DESCRIPTION
**Pull request by DonIsaac**
Re-creation of #5329

### CODE REVIEW

**Comment by DonIsaac** [comment]
* **#5685** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5685?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: apps/oxlint/src/lint/mod.rs [hunk 1]
```diff
@@ -184,6 +190,44 @@ impl LintRunner {
         }
         diagnostic_service
     }
+
+    // moved into a separate function for readability, but it's only ever used
+    // in one place.
+    #[inline]
```
Comment (Boshen) [inline apps/oxlint/src/lint/mod.rs:196]: Probably don't need to add these kind of optimization. Trust the compiler (LLVM) ;-)

File: crates/oxc_linter/src/options/filter.rs [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+/// Enables, disables, and sets the severity of lint rules.
+///
+/// Filters come in 3 forms:
+/// 1. Filter by rule name and/or plugin: `no-const-assign`, `eslint/no-const-assign`
+/// 2. Filter an entire category: `correctness`
+/// 3. Some unknow filter. This is a fallback used when parsing a filter string,
+///    and is interpreted uniquely by the linter.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct LintFilter(AllowWarnDeny, LintFilterKind);
```
Comment (Boshen) [inline crates/oxc_linter/src/options/filter.rs:15]: I dislike tuple syntax, the next person need to convert it to `{}` syntax if a field is added.

This also creates a confusing `.0` `.1` below.

File: crates/oxc_linter/src/options/filter.rs [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+impl LintFilter {
```
Comment (Boshen) [inline crates/oxc_linter/src/options/filter.rs:16]: Can you space out (add a newline) to these lines.

File: crates/oxc_linter/src/options/filter.rs [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+        Self(AllowWarnDeny::Warn, LintFilterKind::Category(RuleCategory::Correctness))
+    }
+}
+
+impl From<LintFilter> for (AllowWarnDeny, LintFilterKind) {
+    fn from(val: LintFilter) -> Self {
+        (val.0, val.1)
+    }
+}
+impl<'a> From<&'a LintFilter> for (AllowWarnDeny, &'a LintFilterKind) {
```
Comment (Boshen) [inline crates/oxc_linter/src/options/filter.rs:65]: Need a newline.

File: crates/oxc_linter/src/options/filter.rs [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+impl From<RuleCategory> for LintFilterKind {
+    #[inline]
+    fn from(category: RuleCategory) -> Self {
+        LintFilterKind::Category(category)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum InvalidFilterKind {
```
Comment (Boshen) [inline crates/oxc_linter/src/options/filter.rs:180]: We can use `thiserror`.

**Review by Boshen — CHANGES_REQUESTED** [review]
👍 Very Rusty!

File: crates/oxc_linter/src/options/filter.rs (hunk 5 above)
Comment (DonIsaac) [reply crates/oxc_linter/src/options/filter.rs:180]: Both `miette` and `thiserror` were removed from the linter's dependencies in favor of `oxc_diagnostics`. This crate does not re-export anything from `thiserror`. Are you OK with me exporting `thiserror`'s macros from this crate?

Comment (Boshen) [reply crates/oxc_linter/src/options/filter.rs:188]: Oh ... I removed `thiserror` everywhere.

**Review by Boshen — APPROVED** [review]

**Pull request merged**